zookeeper-api/src/main/java/org/apache/helix/zookeeper/util/HttpRoutingDataReader.java

```diff
@@ -0,0 +1,134 @@
+import org.apache.http.HttpEntity;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.codehaus.jackson.map.ObjectMapper;
+
+
+public class HttpRoutingDataReader {
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+  private static volatile Map<String, List<String>> _rawRoutingData;
+  private static volatile MetadataStoreRoutingData _metadataStoreRoutingData;
+
+  /**
+   * This class is a Singleton.
+   */
+  private HttpRoutingDataReader() {
+  }
+
+  /**
+   * Fetches routing data from the data source via HTTP.
+   * @return a mapping from "metadata store realm addresses" to lists of
+   * "metadata store sharding keys", where the sharding keys in a value list all route to
+   * the realm address in the key disallows a meaningful mapping to be returned
+   */
+  @RetryOnFailure(attempts = 3, delay = 1, unit = TimeUnit.SECONDS)
+  public static Map<String, List<String>> getRawRoutingData()
+      throws IOException {
+    if (_rawRoutingData == null) {
+      synchronized (HttpRoutingDataReader.class) {
+        if (_rawRoutingData == null) {
+          String msdsEndpoint =
+              System.getProperty(MetadataStoreRoutingConstants.MSDS_SERVER_ENDPOINT_KEY);
+          if (msdsEndpoint == null || msdsEndpoint.isEmpty()) {
+            throw new IllegalStateException(
+                "HttpRoutingDataReader was unable to find a valid MSDS endpoint String in System Properties!");
+          }
+
+          // Note: HttpClient's timeout settings are system timeout settings by default
+          CloseableHttpClient httpClient = HttpClients.createDefault();
+          // Update the reference
+          _rawRoutingData = getAllRoutingData(httpClient, msdsEndpoint);
+          // Close any open resources
+          httpClient.close();
+        }
+      }
+    }
+    return _rawRoutingData;
+  }
+
+  /**
+   * Returns the routing data read from MSDS in a MetadataStoreRoutingData format.
+   * @return
+   * @throws IOException
+   * @throws InvalidRoutingDataException
+   */
+  public static MetadataStoreRoutingData getMetadataStoreRoutingData()
+      throws IOException, InvalidRoutingDataException {
+    if (_metadataStoreRoutingData == null) {
+      synchronized (HttpRoutingDataReader.class) {
+        if (_metadataStoreRoutingData == null) {
+          _metadataStoreRoutingData = new TrieRoutingData(getRawRoutingData());
+        }
+      }
+    }
+    return _metadataStoreRoutingData;
+  }
+
+  /**
+   * Makes an HTTP call to fetch all routing data.
+   * @param httpClient
+   * @param msdsEndpoint
+   * @return
+   * @throws IOException
+   */
+  private static Map<String, List<String>> getAllRoutingData(CloseableHttpClient httpClient,
```

Comment:
Maybe I would decouple fetch http json response and parse json to map:
- response = httpClient.execute(routing-data-endpoint) // Easier to just retry and timeout http request.
- Map<String, List<>> parseRoutingData(response) {}

Reply:
I agree. Let me split the logic into two.